### DISCUSSION TITLE
Mesh position slow to change when updated in animateReady

### DISCUSSION THREAD

**Discussion by IRobot1** [nartc/angular-three#65] [Q&A] [answered]
I have a simple mesh who's position I'm changing in animateReady

```html
<ngt-mesh ngtPhysicBox [getPhysicProps]="getCubeProps" 
          [position]="position" (animateReady)="animateGroup($event.state)">
  <ngt-sphere-geometry [args]="[ radius, 16, 16]"></ngt-sphere-geometry>
  <ngt-mesh-standard-material [parameters]="{ wireframe: true, color: 'black' }"></ngt-mesh-standard-material>
</ngt-mesh>
```

The position is being changed to match the position of the XR motion controller is world space.  

```javascript
  animateReady(event: NgtRender) {
    if (this.controller) {
      const position = this.controller.position;
      this.position = [position.x, position.y, position.z] as NgtVector3;
    }
  }
```

However, the mesh only seems to update twice a second and is slow to track the position of the controller.

![slow physics](https://user-images.githubusercontent.com/25032599/155165138-790c7b6e-6429-4906-9f7d-61eb235fa14f.png)

I've confirmed animateReady function is being called 60 times a second.

Any thoughts?

If you have a VR headset, you can try the project [here](https://github.com/IRobot1/ng3-cannon-template).  Set vr=true in app.component.ts first.

**Comment by nartc** [comment]
Can you replicate this issue without vr mode? I have a hunch of what's going on but it'd be easier with a sample that I can run :)

**Comment by IRobot1** [comment, marked as answer]
I was able to fix by adding change detection

```javascript
  constructor(
    private cd: ChangeDetectorRef,
  ) { }
  animateGroup(event: NgtRender) {
    if (this.controller) {
      this.position = this.controller.position.toArray();
      this.rotation = this.controller.rotation;
      this.cd.detectChanges();
    }
  }
```

**Reply by nartc** [reply]
That really is not a good solution. You’re triggering Change Detection at least 60 times per second and that is a lot. If you can provide a reproduce without vr, I can take a closer look

**Reply by IRobot1** [reply]
Ok.  I'll try to find a way with VR, but it might not be possible.

**Reply by IRobot1** [reply]
I have confirmed that this problem only happens when physics is added to the mesh.  I'm trying to create a simple repeat case

**Reply by IRobot1** [reply]
Ignore that last comment. Its still sluggish with physics removed.
I've even tried generating an event so the change is made from outside the component.  Still slow.  I don't get it.
I'm going to see if I can repeat by following the mouse.

**Reply by nartc** [reply]
https://angular-three.netlify.app/?path=/story/cannon-object-clump--default this example updates the mouse position and such. Have you taken a look?

**Reply by IRobot1** [reply]
The solution I found was to remove the animateReady and put into a setInterval timer.  For some reason, this works.  Running at 90 frames a second avoid any stuttering.

```javascript
  controller?: Group;

  ngOnInit(): void {
    ...
    setInterval(() => {
      if (this.controller) {
        this.position = this.controller.position.toArray();
        this.rotation = this.controller.rotation;
      }
    }, 1000/90)
  }
```

**Reply by nartc** [reply]
Interesting! Now I really want to see the code to find out the root cause.

**Reply by IRobot1** [reply]
Regarding your new object clumping example, I think the PointerComponent needs some documentation and clarification.  It appears to use internal features.  I'd like to understand your thinking on implementing this way.

When to extend from NgtStore and why?
When to use animation frame store?
What is this.effect?
What is tapEffect?
Why do all your components have ChangeDetectionStrategy.OnPush?

**Reply by IRobot1** [reply]
The project code is [here](https://github.com/IRobot1/ng3-cannon-template).  See xr-controller.component.ts.  Look at history to see version that didn't work.

I've updated position in other animationReady callbacks, and they've all worked fine (as you would expect).  Very puzzling.

**Reply by IRobot1** [reply]
The object clump examples was very helpful.  

Changing the location and rotation in the animateReady callback using physics api works great!  I've removed my work-around mentioned above.

```javascript
  animate() {
    const p = this.controller.position;
    this.physics.api.position.set(p.x, p.y, p.z);

    const r = this.controller.rotation;
    this.physics.api.rotation.set(r.x, r.y, r.z);
  }
```
```html
<ngt-mesh ngtPhysicBox [getPhysicProps]="getCubeProps" (animateReady)="animate()"
```

This has also solved a problem with collideBegin and collideEnd events not firing properly when one object uses isTrigger.

Thanks for this useful example.

**Reply by nartc** [reply]
That is great to hear!!

About those clarification questions, I agree. I will update the docs when I find some time.
